lsr-role2collection.py

```diff
@@ -0,0 +1,764 @@
+                with open(filepath) as f:
+                    s = f.read()
+                s = re.sub(find, replace, s)
+                with open(filepath, "w") as f:
+                    f.write(s)
+
+
+def replace_rolepath_with_collection(namespace, collection, role):
+    """
+    Replace "{{ role_path }}/roles/rolename" with namespace.collection.rolename
+    in role_dir.
+    """
+    role_dir = output / 'roles' / role
+    find = "{{ role_path }}/roles/(.*)"
+    replace = namespace + "." + collection + ".\\1"
+    file_patterns = ['*.yml', '*.md']
+    file_replace(role_dir, find, replace, file_patterns)
+
+# Replace "{{ role_path }}/roles/ROLE" with NAMESPACE.COLLECTION.ROLE
+replace_rolepath_with_collection(namespace, collection, role)
+
+# ==============================================================================
+
+# Tests - copy files and dirs in the tests to
+# DEST_PATH/ansible_collections/NAMESPACE/COLLECTION/tests/ROLE.
+def copy_tests(src_path, role):
+    for tests in TESTS:
+        src = src_path / tests
+        if src.is_dir():
+            dest = output / tests / role
+            print(f'Copying role {src} to {dest}')
+            copytree(
+                src,
+                dest,
+                ignore=ignore_patterns('artifacts'),
+                symlinks=True,
+                dirs_exist_ok=True
+            )
+
+
+# Adjust role names to the collections style.
+def remove_or_reset_symlinks(path, role):
+    """
+    Clean up roles/linux-system-roles.rolename.
+    - Remove symlinks.
+    - If linux-system-roles.rolename is an empty dir, rmdir it.
+    """
+    nodes = sorted(list(path.rglob('*')), reverse=True)
+    for node in nodes:
+        if node.is_symlink():
+            node.unlink()
+        elif node.is_dir() and r'linux-system-roles.' + role == node.name and not any(node.iterdir()):
+            node.rmdir()
+
+
+def recursive_grep(path, find, file_patterns):
+    """
+    Check if a pattern `find` is found in the files that match
+    `file_patterns` under `path`.
+    """
+    for root, dirs, files in os.walk(os.path.abspath(path)):
+        for file_pattern in file_patterns:
+            for filename in fnmatch.filter(files, file_pattern):
+                filepath = os.path.join(root, filename)
+                with open(filepath) as f:
+                    s = f.read()
+                if find in s:
+                    return True
+    return False
+
+
+def replace_rolename_with_collection(path, role, collection):
+    """
+    Replace the roles or include_role values, ROLE or `linux-system-roles.ROLE`, are replaced
+    with `NAMESPACE.COLLECTION.ROLE` in the given dir `path` recursively.
+    """
+    find = r"( *name: | *- name: | *- | *roletoinclude: | *role: | *- role: )(linux-system-roles\.{0}\b|{0}\b)".format(role, role)
+    replace = r"\1" + namespace + "." + collection + "." + role
+    file_patterns = ['*.yml', '*.md']
+    file_replace(path, find, replace, file_patterns)
```

Comment:
This doesn't work with the network role.  For example: https://github.com/linux-system-roles/network/blob/main/README.md
```yaml
      route:
        - network: 198.51.100.128
          prefix: 26
...
```
is converted to
```yaml
      route:
        - fedora.system_roles.network: 198.51.100.128
          prefix: 26
```
I think we should enforce that every place where the role name is mentioned, it is mentioned using the fully qualified name, and this script should only have to convert references to `linux-system-roles.ROLENAME` to `NAMESPACE.system_roles.ROLENAME`.  Would that simplify the script?

Reply:
@richm, that'd be ideal. Indeed, in addition to solving the bug you found, I can get rid of `|{0}\b` from the pattern!

Just out of my curiosity, this is the snippet of the python doc. I was hoping I could use ^ and $ in the pattern. Although the doc says it should work in MULTIPLE mode, it did not solve my problem... (or may be I could use \n instead a in [0] to avoid a problem like in the network role?)

$
    Matches the end of the string or just before the newline at the end of the string, and in MULTILINE mode also matches before a newline. foo matches both ‘foo’ and ‘foobar’, while the regular expression foo$ matches only ‘foo’. More interestingly, searching for foo.$ in 'foo1\nfoo2\n' matches ‘foo2’ normally, but ‘foo1’ in MULTILINE mode; searching for a single $ in 'foo\n' will find two (empty) matches: one just before the newline, and one at the end of the string.

[0] - find = r"( *name: | *- name: | *- | *roletoinclude: | *role: | *- role: )(linux-system-roles\.{0}\n|{0}\n)".format(role, role)

Reply:
> This doesn't work with the network role. For example: https://github.com/linux-system-roles/network/blob/main/README.md
> 
> ```yaml
>       route:
>         - network: 198.51.100.128
>           prefix: 26
> ...
> ```
> 
> is converted to (...)

This occurrence of `network` is not in a place where one expect a role name, so the tool should not even attempt to convert this. Which highlights the importance of using a parser that properly understand the YAML structure.